src/vm/i386/umthunkstub.S

```diff
@@ -11,22 +11,132 @@
 //
 NESTED_ENTRY TheUMEntryPrestub, _TEXT, UnhandledExceptionHandlerUnix
     // Preserve argument registers
-    push    ecx
+    push    ebx
     push    edx
 
-    push    eax
+    push    eax  // UMEntryThunk*
     call    C_FUNC(TheUMEntryPrestubWorker)
     pop     edx
+    // eax = PCODE
 
     // Restore argument registers
     pop     edx
-    pop     ecx
+    pop     ebx
 
     jmp     eax     // Tail Jmp
 NESTED_END TheUMEntryPrestub, _TEXT
 
 
+#define UMThunkStub_FIXEDALLOCSIZE          0x20
+#define UMThunkStub_UMENTRYTHUNK_OFFSET     0x04
+#define UMThunkStub_THREAD_OFFSET           0x08
+#define UMThunkStub_INT_ARG_OFFSET          0x00
+#define UMThunkStub_EBP_OFFSET              (UMThunkStub_FIXEDALLOCSIZE)
+
+
 NESTED_ENTRY UMThunkStub, _TEXT, UnhandledExceptionHandlerUnix
-    int     3       // implement here
+    PROLOG_BEG
+    PROLOG_END
+    sub     esp, UMThunkStub_FIXEDALLOCSIZE + 4
```

Comment:
It would be better to incorporate the "+4" in the UMThunkStub_FIXEDALLOCSIZE instead to reflect the fact that that constant is the amount of memory we have allocated on the stack.